File: apps/platform-web/src/services/requirement-review/requirement-review.service.ts
import { platformHttpClient } from '@/services/http/client'
import {
  downloadOperationArtifact,
  submitOperation,
  waitForOperationTerminalState
} from '@/services/operations/operations.service'
import type { ChatAttachmentBlock } from '@/utils/chat-content'
import type {
  ManagementDownload,
  ManagementOperation,
  RequirementFeatureList,
  RequirementReviewBatchDetail,
  RequirementReviewBatchSummary,
  RequirementReviewDocument,
  RequirementReviewOverview,
  RequirementReviewRole,
  RequirementReviewResult
} from '@/types/management'

type BatchListResponse = {
  items: RequirementReviewBatchSummary[]
  total: number
}

type DocumentListResponse = {
  items: RequirementReviewDocument[]
  total: number
}

type ResultListResponse = {
  items: RequirementReviewResult[]
  total: number
}

export type UpsertRequirementReviewDocumentPayload = {
  batch_id?: string | null
  thread_id?: string | null
  filename?: string
  content_type?: string
  source_kind?: string
  parse_status?: string
  summary_for_model?: string
  parsed_text?: string | null
  structured_data?: Record<string, unknown> | null
  provenance?: Record<string, unknown> | null
  error?: Record<string, unknown> | null
}

export type UpsertRequirementReviewResultPayload = {
  batch_id?: string | null
  thread_id?: string | null
  document_ids?: string[] | null
  requirement_summary?: string
  review_score?: number | null
  quality_gate?: string
  dimension_scores?: Record<string, unknown> | null
  key_findings?: string[] | null
  major_risks?: string[] | null
  missing_or_ambiguous_items?: string[] | null
  suggestions_to_improve?: string[] | null
  generation_policy?: string
  generation_policy_reason?: string
  assumptions?: string[] | null
  raw_result?: Record<string, unknown> | null
}

function buildHeaders(projectId: string) {
  return {
    'x-project-id': projectId
  }
}

function resolveEndpoint(projectId: string, suffix: string) {
  const normalizedSuffix = suffix.startsWith('/') ? suffix : `/${suffix}`
  return {
    client: platformHttpClient,
    path: `/api/projects/${projectId}/requirement-review${normalizedSuffix}`
  }
}

function parseContentDispositionFilename(header: string | null): string | null {
  if (!header) {
    return null
  }
  const utf8Match = header.match(/filename\*=UTF-8''([^;]+)/i)
  if (utf8Match?.[1]) {
    try {
      return decodeURIComponent(utf8Match[1])
    } catch {
      return utf8Match[1]
    }
  }
  const plainMatch = header.match(/filename="?([^";]+)"?/i)
  return plainMatch?.[1] ?? null
}

async function downloadFrom(
  projectId: string,
  suffix: string,
  params?: Record<string, string | number | undefined>
): Promise<ManagementDownload> {
  const { client, path } = resolveEndpoint(projectId, suffix)
  const response = await client.get(path, {
    headers: buildHeaders(projectId),
    responseType: 'blob',
    params
  })
  return {
    blob: response.data as Blob,
    filename: parseContentDispositionFilename(String(response.headers['content-disposition'] || '')),
    contentType: String(response.headers['content-type'] || '')
  }
}

export async function getRequirementReviewOverview(
  projectId: string
): Promise<RequirementReviewOverview> {
  const { client, path } = resolveEndpoint(projectId, '/overview')
  const response = await client.get(path, {
    headers: buildHeaders(projectId)
  })
  return response.data as RequirementReviewOverview
}

export async function getRequirementReviewRole(
  projectId: string
): Promise<RequirementReviewRole> {
  const { client, path } = resolveEndpoint(projectId, '/role')
  const response = await client.get(path, {
    headers: buildHeaders(projectId)
  })
  return response.data as RequirementReviewRole
}

export async function listRequirementReviewBatches(
  projectId: string,
  options?: { limit?: number; offset?: number }
): Promise<BatchListResponse> {
  const { client, path } = resolveEndpoint(projectId, '/batches')
  const response = await client.get(path, {
    headers: buildHeaders(projectId),
    params: {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0
    }
  })
  return response.data as BatchListResponse
}

export async function getRequirementReviewBatchDetail(
  projectId: string,
  batchId: string,
  options?: {
    document_limit?: number
    document_offset?: number
    result_limit?: number
    result_offset?: number
  }
): Promise<RequirementReviewBatchDetail> {
  const { client, path } = resolveEndpoint(projectId, `/batches/${batchId}`)
  const response = await client.get(path, {
    headers: buildHeaders(projectId),
    params: {
      document_limit: options?.document_limit ?? 100,
      document_offset: options?.document_offset ?? 0,
      result_limit: options?.result_limit ?? 50,
      result_offset: options?.result_offset ?? 0
    }
  })
  return response.data as RequirementReviewBatchDetail
}

export async function listRequirementReviewDocuments(
  projectId: string,
  options?: {
    batch_id?: string
    parse_status?: string
    query?: string
    limit?: number
    offset?: number
  }
): Promise<DocumentListResponse> {
  const { client, path } = resolveEndpoint(projectId, '/documents')
  const response = await client.get(path, {
    headers: buildHeaders(projectId),
    params: {
      limit: options?.limit ?? 20,
      offset: options?.offset ?? 0,
      batch_id: options?.batch_id?.trim() || undefined,
      parse_status: options?.parse_status?.trim() || undefined,
      query: options?.query?.trim() || undefined
    }
  })
  return response.data as DocumentListResponse
}

export async function getRequirementReviewDocument(
  projectId: string,
  documentId: string
): Promise<RequirementReviewDocument> {
  const { client, path } = resolveEndpoint(projectId, `/documents/${documentId}`)
  const response = await client.get(path, {
    headers: buildHeaders(projectId)
  })
  return response.data as RequirementReviewDocument
}

export async function previewRequirementReviewDocument(
  projectId: string,
  documentId: string
): Promise<ManagementDownload> {
  return downloadFrom(projectId, `/documents/${documentId}/preview`)
}

export async function downloadRequirementReviewDocument(
  projectId: string,
  documentId: string
): Promise<ManagementDownload> {
  return downloadFrom(projectId, `/documents/${documentId}/download`)
}

export async function exportRequirementReviewDocuments(
  projectId: string,
  options?: {
    batch_id?: string
    parse_status?: string
    query?: string
  }
): Promise<ManagementDownload> {
  return downloadFrom(projectId, '/documents/export', {
    batch_id: options?.batch_id?.trim() || undefined,
    parse_status: options?.parse_status?.trim() || undefined,
    query: options?.query?.trim() || undefined
  })
}

export async function exportRequirementReviewDocumentsByOperation(
  projectId: string,
  options?: {
    batch_id?: string
    parse_status?: string
    query?: string
    idempotencyKey?: string
  }
): Promise<{
  operation: ManagementOperation
  download: ManagementDownload
}> {
  const submitted = await submitOperation({
    kind: 'requirement-review.documents.export',
    project_id: projectId,
    idempotency_key: options?.idempotencyKey,
    input_payload: {
      batch_id: options?.batch_id?.trim() || undefined,
      parse_status: options?.parse_status?.trim() || undefined,
      query: options?.query?.trim() || undefined
    }
  })
  const operation = await waitForOperationTerminalState(submitted.id, {
    pollMs: 1000,
    timeoutMs: 120000
  })
  const download = await downloadOperationArtifact(operation.id)
  return {
    operation,
    download
  }
}

export async function createRequirementReviewDocument(
  projectId: string,
  payload: Required<Pick<UpsertRequirementReviewDocumentPayload, 'filename' | 'content_type'>> &
    UpsertRequirementReviewDocumentPayload
): Promise<RequirementReviewDocument> {
  const { client, path } = resolveEndpoint(projectId, '/documents')
  const response = await client.post(path, payload, {
    headers: buildHeaders(projectId)
  })
  return response.data as RequirementReviewDocument
}

export async function updateRequirementReviewDocument(
  projectId: string,
  documentId: string,
  payload: UpsertRequirementReviewDocumentPayload
): Promise<RequirementReviewDocument> {
  const { client, path } = resolveEndpoint(projectId, `/documents/${documentId}`)
  const response = await client.patch(path, payload, {
    headers: buildHeaders(projectId)
  })
  return response.data as RequirementReviewDocument
}

export async function deleteRequirementReviewDocument(
  projectId: string,
  documentId: string
): Promise<void> {
  const { client, path } = resolveEndpoint(projectId, `/documents/${documentId}`)
  await client.delete(path, {
    headers: buildHeaders(projectId)
  })
}

export async function listRequirementReviewResults(
  projectId: string,
  options?: {
    batch_id?: string
    quality_gate?: string
    generation_policy?: string
    query?: string
    limit?: number
    offset?: number
  }
): Promise<ResultListResponse> {
  const { client, path } = resolveEndpoint(projectId, '/results')
  const response = await client.get(path, {
    headers: buildHeaders(projectId),
    params: {
      limit: options?.limit ?? 20,
      offset: options?.offset ?? 0,
      batch_id: options?.batch_id?.trim() || undefined,
      quality_gate: options?.quality_gate?.trim() || undefined,
      generation_policy: options?.generation_policy?.trim() || undefined,
      query: options?.query?.trim() || undefined
    }
  })
  return response.data as ResultListResponse
}

export async function getRequirementReviewResult(
  projectId: string,
  resultId: string
): Promise<RequirementReviewResult> {
  const { client, path } = resolveEndpoint(projectId, `/results/${resultId}`)
  const response = await client.get(path, {
    headers: buildHeaders(projectId)
  })
  return response.data as RequirementReviewResult
}

export async function exportRequirementReviewResults(
  projectId: string,
  options?: {
    batch_id?: string
    quality_gate?: string
    generation_policy?: string
    query?: string
  }
): Promise<ManagementDownload> {
  return downloadFrom(projectId, '/results/export', {
    batch_id: options?.batch_id?.trim() || undefined,
    quality_gate: options?.quality_gate?.trim() || undefined,
    generation_policy: options?.generation_policy?.trim() || undefined,
    query: options?.query?.trim() || undefined
  })
}

export async function exportRequirementReviewResultsByOperation(
  projectId: string,
  options?: {
    batch_id?: string
    quality_gate?: string
    generation_policy?: string
    query?: string
    idempotencyKey?: string
  }
): Promise<{
  operation: ManagementOperation
  download: ManagementDownload
}> {
  const submitted = await submitOperation({
    kind: 'requirement-review.results.export',
    project_id: projectId,
    idempotency_key: options?.idempotencyKey,
    input_payload: {
      batch_id: options?.batch_id?.trim() || undefined,
      quality_gate: options?.quality_gate?.trim() || undefined,
      generation_policy: options?.generation_policy?.trim() || undefined,
      query: options?.query?.trim() || undefined
    }
  })
  const operation = await waitForOperationTerminalState(submitted.id, {
    pollMs: 1000,
    timeoutMs: 120000
  })
  const download = await downloadOperationArtifact(operation.id)
  return {
    operation,
    download
  }
}

export async function createRequirementReviewResult(
  projectId: string,
  payload: Required<Pick<UpsertRequirementReviewResultPayload, 'generation_policy'>> &
    UpsertRequirementReviewResultPayload
): Promise<RequirementReviewResult> {
  const { client, path } = resolveEndpoint(projectId, '/results')
  const response = await client.post(path, payload, {
    headers: buildHeaders(projectId)
  })
  return response.data as RequirementReviewResult
}

export async function updateRequirementReviewResult(
  projectId: string,
  resultId: string,
  payload: UpsertRequirementReviewResultPayload
): Promise<RequirementReviewResult> {
  const { client, path } = resolveEndpoint(projectId, `/results/${resultId}`)
  const response = await client.patch(path, payload, {
    headers: buildHeaders(projectId)
  })
  return response.data as RequirementReviewResult
}

export async function deleteRequirementReviewResult(
  projectId: string,
  resultId: string
): Promise<void> {
  const { client, path } = resolveEndpoint(projectId, `/results/${resultId}`)
  await client.delete(path, {
    headers: buildHeaders(projectId)
  })
}

type FeatureListListResponse = {
  items: RequirementFeatureList[]
  total: number
}

export async function listRequirementFeatureLists(
  projectId: string,
  options?: {
    batch_id?: string
    status?: string
    query?: string
    limit?: number
    offset?: number
  }
): Promise<FeatureListListResponse> {
  const { client, path } = resolveEndpoint(projectId, '/feature-lists')
  const response = await client.get(path, {
    headers: buildHeaders(projectId),
    params: {
      limit: options?.limit ?? 20,
      offset: options?.offset ?? 0,
      batch_id: options?.batch_id?.trim() || undefined,
      status: options?.status?.trim() || undefined,
      query: options?.query?.trim() || undefined
    }
  })
  return response.data as FeatureListListResponse
}

export async function getRequirementFeatureList(
  projectId: string,
  featureListId: string
): Promise<RequirementFeatureList> {
  const { client, path } = resolveEndpoint(projectId, `/feature-lists/${featureListId}`)
  const response = await client.get(path, {
    headers: buildHeaders(projectId)
  })
  return response.data as RequirementFeatureList
}

export type UpdateRequirementFeatureListPayload = {
  requirement_text?: string
  requirement_summary?: string
  modules?: Record<string, unknown>[]
  open_questions?: string[]
  assumptions?: string[]
}

export async function updateRequirementFeatureList(
  projectId: string,
  featureListId: string,
  payload: UpdateRequirementFeatureListPayload
): Promise<RequirementFeatureList> {
  const { client, path } = resolveEndpoint(projectId, `/feature-lists/${featureListId}`)
  const response = await client.patch(path, payload, {
    headers: buildHeaders(projectId)
  })
  return response.data as RequirementFeatureList
}

export async function confirmRequirementFeatureList(
  projectId: string,
  featureListId: string,
  options?: { expected_version?: number }
): Promise<RequirementFeatureList> {
  const { client, path } = resolveEndpoint(
    projectId,
    `/feature-lists/${featureListId}/confirm`
  )
  const response = await client.post(
    path,
    { expected_version: options?.expected_version },
    { headers: buildHeaders(projectId) }
  )
  return response.data as RequirementFeatureList
}

export async function deleteRequirementFeatureList(
  projectId: string,
  featureListId: string
): Promise<void> {
  const { client, path } = resolveEndpoint(projectId, `/feature-lists/${featureListId}`)
  await client.delete(path, {
    headers: buildHeaders(projectId)
  })
}

export async function decomposeRequirementByOperation(
  projectId: string,
  options: {
    requirement_text?: string
    attachments?: ChatAttachmentBlock[]
    batch_id?: string
    idempotencyKey?: string
  }
): Promise<ManagementOperation> {
  const submitted = await submitOperation({
    kind: 'requirement.feature_list.decompose',
    project_id: projectId,
    idempotency_key: options.idempotencyKey,
    input_payload: {
      requirement_text: options.requirement_text?.trim() || undefined,
      attachments: options.attachments?.length ? options.attachments : undefined,
      batch_id: options.batch_id?.trim() || undefined
    }
  })
  return waitForOperationTerminalState(submitted.id, {
    pollMs: 2000,
    timeoutMs: 300000
  })
}

export async function reviewAndGenerateByOperation(
  projectId: string,
  options: {
    feature_list_id?: string
    requirement_text?: string
    idempotencyKey?: string
  }
): Promise<ManagementOperation> {
  const submitted = await submitOperation({
    kind: 'testcase.review_and_generate',
    project_id: projectId,
    idempotency_key: options.idempotencyKey,
    input_payload: {
      feature_list_id: options.feature_list_id || undefined,
      requirement_text: options.requirement_text?.trim() || undefined
    }
  })
  return waitForOperationTerminalState(submitted.id, {
    pollMs: 2000,
    timeoutMs: 600000
  })
}

File: apps/interaction-data-service/app/schemas/requirement_review_service.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class CreateRequirementReviewDocumentRequest(BaseModel):
    project_id: str
    batch_id: str | None = None
    thread_id: str | None = Field(default=None, max_length=255)
    idempotency_key: str | None = Field(default=None, max_length=255)
    filename: str = Field(min_length=1, max_length=255)
    content_type: str = Field(min_length=1, max_length=128)
    storage_path: str | None = None
    source_kind: str = Field(default="upload", min_length=1, max_length=64)
    parse_status: str = Field(default="parsed", min_length=1, max_length=64)
    summary_for_model: str = ""
    parsed_text: str | None = None
    structured_data: dict[str, Any] | None = None
    provenance: dict[str, Any] = Field(default_factory=dict)
    error: dict[str, Any] | None = None


class RequirementReviewDocumentResponse(BaseModel):
    id: str
    project_id: str
    batch_id: str | None
    thread_id: str | None
    idempotency_key: str | None
    filename: str
    content_type: str
    storage_path: str | None = None
    source_kind: str
    parse_status: str
    summary_for_model: str
    parsed_text: str | None
    structured_data: dict[str, Any] | None
    provenance: dict[str, Any]
    error: dict[str, Any] | None
    created_at: str
    updated_at: str


class RequirementReviewDocumentListResponse(BaseModel):
    items: list[RequirementReviewDocumentResponse]
    total: int


class RequirementReviewDocumentAssetResponse(BaseModel):
    storage_path: str
    filename: str
    content_type: str
    size: int
    sha256: str


class UpdateRequirementReviewDocumentRequest(BaseModel):
    batch_id: str | None = None
    thread_id: str | None = Field(default=None, max_length=255)
    filename: str | None = Field(default=None, min_length=1, max_length=255)
    content_type: str | None = Field(default=None, min_length=1, max_length=128)
    storage_path: str | None = None
    source_kind: str | None = Field(default=None, min_length=1, max_length=64)
    parse_status: str | None = Field(default=None, min_length=1, max_length=64)
    summary_for_model: str | None = None
    parsed_text: str | None = None
    structured_data: dict[str, Any] | None = None
    provenance: dict[str, Any] | None = None
    error: dict[str, Any] | None = None


class CreateRequirementReviewResultRequest(BaseModel):
    project_id: str
    batch_id: str | None = None
    thread_id: str | None = Field(default=None, max_length=255)
    idempotency_key: str | None = Field(default=None, max_length=255)
    document_ids: list[str] = Field(default_factory=list)
    requirement_summary: str = ""
    review_score: float | None = Field(default=None, ge=0, le=100)
    quality_gate: str = Field(default="conditional", min_length=1, max_length=64)
    dimension_scores: dict[str, Any] = Field(default_factory=dict)
    key_findings: list[str] = Field(default_factory=list)
    major_risks: list[str] = Field(default_factory=list)
    missing_or_ambiguous_items: list[str] = Field(default_factory=list)
    suggestions_to_improve: list[str] = Field(default_factory=list)
    generation_policy: str = Field(min_length=1, max_length=128)
    generation_policy_reason: str = ""
    assumptions: list[str] = Field(default_factory=list)
    raw_result: dict[str, Any] = Field(default_factory=dict)


class RequirementReviewResultResponse(BaseModel):
    id: str
    project_id: str
    batch_id: str | None
    thread_id: str | None
    idempotency_key: str | None
    document_ids: list[str]
    requirement_summary: str
    review_score: float | None
    quality_gate: str
    dimension_scores: dict[str, Any]
    key_findings: list[str]
    major_risks: list[str]
    missing_or_ambiguous_items: list[str]
    suggestions_to_improve: list[str]
    generation_policy: str
    generation_policy_reason: str
    assumptions: list[str]
    raw_result: dict[str, Any]
    created_at: str
    updated_at: str


class RequirementReviewResultListResponse(BaseModel):
    items: list[RequirementReviewResultResponse]
    total: int


class UpdateRequirementReviewResultRequest(BaseModel):
    batch_id: str | None = None
    thread_id: str | None = Field(default=None, max_length=255)
    document_ids: list[str] | None = None
    requirement_summary: str | None = None
    review_score: float | None = Field(default=None, ge=0, le=100)
    quality_gate: str | None = Field(default=None, min_length=1, max_length=64)
    dimension_scores: dict[str, Any] | None = None
    key_findings: list[str] | None = None
    major_risks: list[str] | None = None
    missing_or_ambiguous_items: list[str] | None = None
    suggestions_to_improve: list[str] | None = None
    generation_policy: str | None = Field(default=None, min_length=1, max_length=128)
    generation_policy_reason: str | None = None
    assumptions: list[str] | None = None
    raw_result: dict[str, Any] | None = None


class CreateRequirementFeatureListRequest(BaseModel):
    project_id: str
    batch_id: str | None = None
    thread_id: str | None = Field(default=None, max_length=255)
    idempotency_key: str | None = Field(default=None, max_length=255)
    decomposable: bool = True
    undecomposable_reason: str | None = None
    requirement_text: str = ""
    requirement_summary: str = ""
    modules: list[dict[str, Any]] = Field(default_factory=list)
    open_questions: list[str] = Field(default_factory=list)
    assumptions: list[str] = Field(default_factory=list)
    raw_result: dict[str, Any] = Field(default_factory=dict)


class UpdateRequirementFeatureListRequest(BaseModel):
    batch_id: str | None = None
    thread_id: str | None = Field(default=None, max_length=255)
    decomposable: bool | None = None
    undecomposable_reason: str | None = None
    requirement_text: str | None = None
    requirement_summary: str | None = None
    modules: list[dict[str, Any]] | None = None
    open_questions: list[str] | None = None
    assumptions: list[str] | None = None
    raw_result: dict[str, Any] | None = None


class ConfirmRequirementFeatureListRequest(BaseModel):
    confirmed_by: str | None = Field(default=None, max_length=255)
    # 乐观校验：调用方声明确认的是哪个版本，不匹配则拒绝
    expected_version: int | None = Field(default=None, ge=1)


class RequirementFeatureListResponse(BaseModel):
    id: str
    project_id: str
    batch_id: str | None
    thread_id: str | None
    idempotency_key: str | None
    version: int
    status: str
    decomposable: bool
    undecomposable_reason: str | None
    requirement_text: str
    requirement_summary: str
    modules: list[dict[str, Any]]
    open_questions: list[str]
    assumptions: list[str]
    raw_result: dict[str, Any]
    confirmed_at: str | None
    confirmed_by: str | None
    created_at: str
    updated_at: str


class RequirementFeatureListListResponse(BaseModel):
    items: list[RequirementFeatureListResponse]
    total: int


class RequirementReviewOverviewResponse(BaseModel):
    project_id: str | None = None
    documents_total: int
    parsed_documents_total: int
    failed_documents_total: int
    results_total: int
    pass_results_total: int
    conditional_results_total: int
    fail_results_total: int
    latest_batch_id: str | None = None
    latest_activity_at: str | None = None


class RequirementReviewBatchSummary(BaseModel):
    batch_id: str
    documents_count: int
    results_count: int
    latest_created_at: str | None = None
    parse_status_summary: dict[str, int] = Field(default_factory=dict)
    quality_gate_summary: dict[str, int] = Field(default_factory=dict)


class RequirementReviewBatchListResponse(BaseModel):
    items: list[RequirementReviewBatchSummary]
    total: int


class RequirementReviewBatchDetailResponse(BaseModel):
    batch: RequirementReviewBatchSummary
    documents: RequirementReviewDocumentListResponse
    results: RequirementReviewResultListResponse
